File: src/dotnet/ForzaStudioCLI/Forza/ForzaResource.cs
using System;
using System.Collections.Generic;
using System.IO;

namespace Forza
{
    public abstract class ForzaResource : IDisposable
    {
        public Guid Id = Guid.NewGuid();

        public Dictionary<string, object> Tags = new Dictionary<string, object>();

        public ForzaVersion Version { get; set; }

        public string FilePath { get; protected set; }

        public string ArchivePath { get; protected set; }

        public string FileName => Path.GetFileNameWithoutExtension(FilePath).ToLowerInvariant();

        public string Extension => Path.GetExtension(FilePath).ToLowerInvariant();

        public string Name { get; set; }

        public string Description { get; set; }

        protected ForzaResource(string filePath, string archivePath = null)
        {
            if (filePath != null)
            {
                FilePath = filePath.ToLowerInvariant();
                if (archivePath != null)
                {
                    ArchivePath = archivePath.ToLowerInvariant();
                }
                return;
            }
            throw new ArgumentNullException("A file path must be specified");
        }

        public byte[] GetData()
        {
            try
            {
                if (!Utilities.IsNullOrWhiteSpace(ArchivePath) && 
                    !Utilities.IsNullOrWhiteSpace(FilePath) && 
                    File.Exists(ArchivePath))
                {
                    using (ForzaArchive forzaArchive = new ForzaArchive(ArchivePath))
                    {
                        // Unable to load DLL 'xcompress.dll' or one of its dependencies: The specified module could not be found.
                        return forzaArchive[FilePath].Data;
                    }
                }
                if (!Utilities.IsNullOrWhiteSpace(FilePath) && File.Exists(FilePath))
                {
                    using (FileStream fileStream = new FileStream(FilePath, FileMode.Open, FileAccess.Read, FileShare.Read))
                    {
                        byte[] array = new byte[fileStream.Length];
                        fileStream.Read(array, 0, array.Length);
                        return array;
                    }
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
            }
            return null;
        }

        public Stream GetDataStream()
        {
            byte[] data = GetData();
            if (data != null)
            {
                return new MemoryStream(data);
            }
            return null;
        }

        public abstract void Dispose();
    }
}

File: src/py/utils/forza_resource.py
import os
import io
import uuid
from typing import Optional
from utils.forza_version import ForzaVersion

class ForzaResource():
    def __init__(self, file_path: str, archive_path: str = None) -> None:
        if file_path is None:
            raise ValueError("A file path must be specified")

        self.id: uuid.UUID = uuid.uuid4()

        self.version: ForzaVersion = ForzaVersion.Unknown
        self._file_path: str = file_path.lower()
        self._archive_path: Optional[str] = archive_path.lower() if archive_path is not None else None

        self.name: Optional[str] = None

    @property
    def file_path(self) -> str:
        return self._file_path

    @property
    def archive_path(self) -> Optional[str]:
        return self._archive_path

    @property
    def file_name(self) -> str:
        base = os.path.basename(self._file_path)
        name, _ext = os.path.splitext(base)
        return name.lower()

    @property
    def extension(self) -> str:
        _base = os.path.basename(self._file_path)
        _name, ext = os.path.splitext(_base)
        return ext.lower()

    def get_data(self) -> bytes:
        with open(self._file_path, "rb") as f:
            return f.read()
    
        return None

    def get_data_stream(self) -> io.BytesIO:
        data = self.get_data()
        return io.BytesIO(data) if data is not None else None